src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,206 @@
 #include <cassert>
 #include <cstdint>
+#include <unordered_map>
 
 #include <pthread.h>
 #include <sched.h>
 
 #include "affinity.h"
+#include "log.h"
 
-namespace celerity {
-namespace detail {
+namespace celerity::detail::thread_pinning {
 
-	uint32_t affinity_cores_available() {
-		cpu_set_t available_cores;
-		[[maybe_unused]] const auto ret = pthread_getaffinity_np(pthread_self(), sizeof(cpu_set_t), &available_cores);
-		assert(ret == 0 && "Error retrieving affinity mask.");
-		return CPU_COUNT(&available_cores);
+struct pinned_thread_state {
+	cpu_set_t previous_cpuset = {};
+	cpu_set_t pinned_to_cpuset = {};
+};
+
+struct thread_pinner_state {
+	std::mutex mutex;
+	bool initialized = false;
+	runtime_configuration config;
+	cpu_set_t available_cores = {};
+	std::unordered_map<thread_type, uint32_t> thread_pinning_plan;
+	std::unordered_map<pthread_t, pinned_thread_state> pinned_threads;
+};
+thread_pinner_state g_state; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// This helper removes a thread from the pinned_threads map when it ends
+struct thread_remover {
+	thread_remover() = default;
+	thread_remover(const thread_remover&) = delete;
+	thread_remover& operator=(const thread_remover&) = delete;
+	thread_remover(thread_remover&&) = delete;
+	thread_remover& operator=(thread_remover&&) = delete;
+	~thread_remover() {
+		std::lock_guard lock(g_state.mutex);
+		if(g_state.initialized) { g_state.pinned_threads.erase(pthread_self()); }
+	}
+};
+thread_local std::optional<thread_remover> t_remover; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// Initializes the thread pinning machinery
+// This captures the current thread's affinity mask and sets the thread pinning machinery up
+// Calls to pin_this_thread prior to this call will have no effect
+void initialize(const runtime_configuration& cfg) {
+	std::lock_guard lock(g_state.mutex);
+	assert(!g_state.initialized && "Thread pinning already initialized.");
+	assert(g_state.thread_pinning_plan.empty() && "Thread pinning plan not initially empty.");
+	assert(g_state.pinned_threads.empty() && "Pinned threads not initially empty.");
+
+	g_state.config = cfg;
+
+	const auto ret = sched_getaffinity(0, sizeof(cpu_set_t), &g_state.available_cores);
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask, will disable pinning.");
+		g_state.config.enabled = false;
+		return;
 	}
 
-} // namespace detail
-} // namespace celerity
+	// pinned threads per process: user, scheduler, executor, 1 backend worker per device
+	uint32_t pinned_threads_per_process = 3 + g_state.config.num_devices;
```

Comment:
Devices should only be enumerated when backend workers / device submission threads are enabled, right?

Reply:
Very good point. Handled+ tested this case.